File: experiments/006-SmalltalkVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260915-110924-39025/analysis/acceptance-project/src/main/java/customerimporter/DataBaseSession.java
package customerimporter;

import java.lang.reflect.Field;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Predicate;

public class DataBaseSession {

    private List<Class<?>> configuration;
    private Map<Class<?>, Set<Object>> tables;
    private int id;

    // instance creation

    public static DataBaseSession forConfiguration(List<Class<?>> aConfiguration) {
        return new DataBaseSession(aConfiguration);
    }

    // initialization

    private DataBaseSession(List<Class<?>> aConfiguration) {
        configuration = aConfiguration;
        tables = new HashMap<>();
        id = 0;
    }

    // transaction management

    public void beginTransaction() {

    }

    public void commit() {
        for (Object aCustomer : tables.getOrDefault(Customer.class, Collections.emptySet())) persistAddressesOf((Customer) aCustomer);
    }

    // closing

    public void close() {

    }

    // persistence - private

    private void defineIdOf(Object anObject) {
        try {
            Field idField = anObject.getClass().getDeclaredField("id");
            idField.setAccessible(true);
            idField.set(anObject, newIdFor(anObject));
        } catch (ReflectiveOperationException aReflectiveOperationException) {
            throw new RuntimeException(aReflectiveOperationException);
        }
    }

    private void delay() {
        try {
            Thread.sleep(100);
        } catch (InterruptedException anInterruptedException) {
            throw new RuntimeException(anInterruptedException);
        }
    }

    @SuppressWarnings("unchecked")
    private <T> Collection<T> objectsOfType(Class<T> aType) {
        return (Collection<T>) tables.getOrDefault(aType, Collections.emptySet());
    }

    private void persistAddressesOf(Customer anObjectWithAddresses) {
        for (Address anAddress : anObjectWithAddresses.addresses()) persist(anAddress);
    }

    // id

    public int newIdFor(Object anObject) {
        id = id + 1;
        return id;
    }

    // persistance

    public void persist(Object anObject) {
        Set<Object> table;

        delay();
        table = tables.computeIfAbsent(anObject.getClass(), aClass -> new HashSet<>());

        defineIdOf(anObject);
        table.add(anObject);

        if (anObject instanceof Customer) persistAddressesOf((Customer) anObject);
    }

    // selecting

    public <T> Collection<T> select(Predicate<T> aCondition, Class<T> aType) {
        delay();
        Collection<T> selected = new HashSet<>();
        for (T anObject : objectsOfType(aType)) {
            if (aCondition.test(anObject)) selected.add(anObject);
        }
        return selected;
    }

    public <T> Collection<T> selectAllOfType(Class<T> aType) {
        delay();
        return new HashSet<>(objectsOfType(aType));
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-203929-73794/output/project/src/customerimporter/data_base_session.py
import time

from customerimporter.customer import Customer


class DataBaseSession:

    _configuration = None
    _tables = None
    _id = None

    # instance creation

    @classmethod
    def for_configuration(cls, a_configuration):
        return cls(a_configuration)

    # initialization

    def __init__(self, a_configuration):
        self._configuration = a_configuration
        self._tables = {}
        self._id = 0

    # transaction management

    def begin_transaction(self):
        pass

    def commit(self):
        for a_customer in self._tables.get(Customer, set()): self._persist_addresses_of(a_customer)

    # closing

    def close(self):
        pass

    # persistence - private

    def _define_id_of(self, an_object):
        setattr(an_object, "_id", self.new_id_for(an_object))

    def _delay(self):
        time.sleep(0.1)

    def _objects_of_type(self, a_type):
        return self._tables.get(a_type, set())

    def _persist_addresses_of(self, an_object_with_addresses):
        for an_address in an_object_with_addresses.addresses(): self.persist(an_address)

    # id

    def new_id_for(self, an_object):
        self._id = self._id + 1
        return self._id

    # persistance

    def persist(self, an_object):
        self._delay()
        table = self._tables.setdefault(type(an_object), set())

        self._define_id_of(an_object)
        table.add(an_object)

        if isinstance(an_object, Customer): self._persist_addresses_of(an_object)

    # selecting

    def select(self, a_condition, a_type):
        self._delay()
        selected = set()
        for an_object in self._objects_of_type(a_type):
            if a_condition(an_object): selected.add(an_object)
        return selected

    def select_all_of_type(self, a_type):
        self._delay()
        return set(self._objects_of_type(a_type))
